lib/cdo/pardot.rb

```diff
@@ -265,12 +268,35 @@ def self.submit_prospect_batch(prospects, config)
       end
     end
 
-    log "Completed Pardot prospect #{config[:operation_name]} batch call. #{num_submitted} submitted, #{prospect_emails.length} succeeded, #{malformed_emails.length} rejected as malformed"
+    prospects_per_second = num_submitted / time_elapsed
+    log "Completed Pardot prospect #{config[:operation_name]} batch call. "\
+        "#{num_submitted} submitted, #{prospect_emails.length} succeeded, "\
+        "#{malformed_emails.length} rejected as malformed. "\
+        "#{time_elapsed.round(1)} secs, #{prospects_per_second.round(1)} "\
+        "prospects/sec"
 
     # Mark Pardot sync time of contacts in our database
     PEGASUS_DB[:contact_rollups].where("email in ?", prospect_emails).update(pardot_sync_at: DateTime.now) unless prospect_emails.empty?
     # Mark any email addresses rejected by Pardot as malformed so we don't keep trying to fruitlessly create them forever
     PEGASUS_DB[:contact_rollups].where("email in ?", malformed_emails).update(email_malformed: true) unless malformed_emails.empty?
+
+    @consecutive_timeout_errors = 0
+
+    # Pardot POST requests sometimes fail with a timeout. This happens
+    # frequently enough that we can't let this be fatal for the entire process.
+    # As long as we don't get too many timeouts in a row, keep going.
+  rescue Net::ReadTimeout => e
+    @consecutive_timeout_errors =  (@consecutive_timeout_errors || 0) + 1
+    log "Pardot API request failed with Net::ReadTimeout "\
+      "(#{@consecutive_timeout_errors} errors)."
+    if @consecutive_timeout_errors >= 3
+      log "Too many consecutive errors, halting the process."
+      raise e
+    end
+
+    log "Continuing process. Prospects from batch with timed out response "\
+        "are in an indeterminate state and will be reconciled on the next "\
+        "process run."
```

Comment:
Naively, something like the following would make more sense (so that we don't needlessly leave stuff in an indeterminate state), assuming the timeout is uncorrelated with the particular request (is it?).
```
retries = 0
begin
  <pardot syncing...>
rescue Net::ReadTimeout => e
  if retries < 3
    log "..."
    retry
  end
  log "..."
  raise e
end
```